control-center/include/etc/confluent/docker/control-center.properties.template

```diff
@@ -1,54 +1,67 @@
-    'CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC': 'confluent.monitoring.interceptor.topic',
-    'CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC_PARTITIONS': 'confluent.monitoring.interceptor.topic.partitions',
-    'CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC_RETENTION_MS': 'confluent.monitoring.interceptor.topic.retention.ms',
-    'CONTROL_CENTER_INTERNAL_TOPICS_PARTITIONS': 'confluent.controlcenter.internal.topics.partitions',
-    'CONTROL_CENTER_INTERNAL_TOPICS_RETENTION_MS': 'confluent.controlcenter.internal.topics.retention.ms',
-    'CONTROL_CENTER_INTERNAL_TOPICS_CHANGELOG_SEGEMENT_BYTES': 'confluent.controlcenter.internal.topics.changelog.segment.bytes',
-    'CONTROL_CENTER_LICENSE': 'confluent.license',
-    'CONTROL_CENTER_CONFLUENT_LICENSE': 'confluent.license',
-    'CONFLUENT_METRICS_TOPIC_PARTITION': 'confluent.metrics.topic.partitions',
-    'CONFLUENT_METRICS_TOPIC_REPLICATION': 'confluent.metrics.topic.replication',
-    'CONFLUENT_SUPPORT_METRICS_ENABLE': 'confluent.support.metrics.enable',
-    'CONFLUENT_METADATA_BOOTSTRAP_SERVER_URLS': 'confluent.metadata.bootstrap.server.urls',
-    'CONFLUENT_METADATA_BASIC_AUTH_USER_INFO':'confluent.metadata.basic.auth.user.info',
-    'PUBLIC_KEY_PATH': 'public.key.path',
-    'CONTROL_CENTER_CPREST_URL': 'cprest.url',
-    'CONFIG_PROVIDERS': 'config.providers',
-    'CONFIG_PROVIDERS_SECUREPASS_CLASS': 'config.providers.securepass.class',
+{% set monitoring_props = {
+    'confluent.monitoring.interceptor.topic': ['CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC'],
+    'confluent.monitoring.interceptor.topic.partitions': ['CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC_PARTITIONS'],
+    'confluent.monitoring.interceptor.topic.retention.ms': ['CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC_RETENTION_MS'],
+    'confluent.monitoring.interceptor.topic.skip.backlog.minutes': ['CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC_SKIP_BACKLOG_MINUTES']
 } -%}
 
-{% set excludes = [] -%}
+{% set other_props = {
+    'confluent.license': ['CONTROL_CENTER_LICENSE', 'CONTROL_CENTER_CONFLUENT_LICENSE'],
+    'public.key.path': ['PUBLIC_KEY_PATH']
+} -%}
 
-{% for k, property in other_props.items() -%}
-{{ excludes.append(k) }}
-{% if env.get(k) != None -%}
-{{property}}={{env[k]}}
+{% macro SET_PROPERTIES(properties, required, excludes) -%}
+{% for property, ks in properties.items() -%}
+{# ENCAPSULATE THE VALUE AS RESULT #}
+{% set ns = namespace(result=None) -%}
+{# FIND THE FIRST NOT NONE VALUE AND SET IT #}
+{% for k in ks -%}
+  {% set _ = excludes.append(k) -%}
+  {% if ns.result == None and env.get(k) != None -%}
+    {% set ns.result = env.get(k) -%}
+  {% endif -%}
+{% endfor -%}
+{# FILL THE TEMPLATE. IF THE PROPERTY IS REQUIRED, SET TO ITS VALUE, OR EMPTY STRING IF NONE. #}
+{# IF THE PROPERTY IS NOT REQUIRED, SET THE VALUE ONLY IF NOT NONE. #}
+{% if required and ns.result != None -%}
+  {{property}}={{ns.result}}
+{% elif required and ns.result == None -%}
+  {{property}}=
```

Comment:
I believe the reason we are including props with empty value is for debuggability? Is there any other reason?

Reply:
Umm, this line is only applicable to the required properties. So before this PR, if no `CONTROL_CENTER_BOOTSTRAP_SERVERS` is set, in the properties file, there will be a line 
```
bootstrap.servers=
```
, which is what line 46 is doing now.

I'm not sure if this is done initially for debuggability, or simply forcing users to specify these properties and not falling back to the default values that c3 provides.

Reply:
With empty it might fallback to default value defined in code. It would be best if we could fail fast in `configure` step itself. Not sure how much extra effort that would be. Feel free to leave it this as-is for now.

Reply:
Just tried with `bootstrap.servers=` in c3 prop file, and c3 failed to start with the below error:
```
Caused by: org.apache.kafka.common.config.ConfigException: No resolvable bootstrap urls given in bootstrap.servers
```
Similarly, for `confluent.controlcenter.command.topic.replication=` in prop file c3 failed to start with the below error:
```
Exception in thread "main" org.apache.kafka.common.config.ConfigException: Invalid value  for configuration confluent.controlcenter.command.topic.replication: Not a number of type SHORT
```

Do you feel like it's enough to wait to fail until the time of c3 startup as above, or fail fast? 

It's quite trivial to change to fail fast so either works, but maybe it's better that c3 handles the failure instead of in the template file. But then again, it would mean that having required properties in the template file would be quite meaningless since c3 is the one that actually handles the failure.

Reply:
@cryoshida checking that certain properties exists here makes sense. For example - if bootstrap.server is missing we know for sure that c3 will fail to start. Hence fail-fast is better user experience. For others, it is best to leave it on C3 to fail or fallback to some default value. Such validation doesn't fall within the purview of this repo.
Also, looks like we already check for certain properties in configure script. Let's leave it to that.


>  For example - if bootstrap.server is missing we know for sure that c3 will fail to start.
For what I said above, a counter argument can be made that `bootstrap.servers` may become optional in the future e.g. kafka-less C3. So such checks should be removed altogether. I think it will be premature to make that change just yet. Let leave the way it is for now.